File: src/check_coreg.pl
#!/usr/bin/perl
use FileHandle;
use File::Basename;
use Getopt::Long;
use File::Copy;

if (($#ARGV + 1) < 2){die <<EOS ;}
*** $0

usage: $0 [options] <post> <SARFILE> 
    <post>    (input)  Posting of the SAR image
    SARFILE   (input)  Input SAR file to process 
    -m <num matches>   Set the number of matches to use (depricated)
    -o <max offset>    Set the maximum allowable offset (meters)
    -e <max error>     Set the maximum allowable standard deviation of offset fit (pixels)

EOS

GetOptions ('m=i' => \$matches_flg,'o=i' => \$offset_flg,'e=f' => \$error_flg );


$post = $ARGV[0];
$arg = $ARGV[1];
($sarfile,$path) = fileparse($arg);

$min_matches = 50;
$max_offset = 50;
$max_error = 2;

print "Checking coregistration for $sarfile using $post meters\n";
 
if ($matches_flg != 0) {
  print "Setting minimum number of matches to $matches_flg\n";
  $min_matches = $matches_flg;
}

if ($offset_flg != 0) {
  print "Setting maximum offset to be $offset_flg\n";
  $max_offset = $offset_flg;
}

if ($error_flg != 0) {
  print "Setting maximum error to be $error_flg\n";
  $max_error = $error_flg;
}

open(LOG,">>coreg_check.log") or die "\nERROR $0: cannot open log file: coreg_check.log\n";
print LOG "SAR file: $sarfile\n";
if ($matches_flg != 0) { print LOG "Checking for $min_matches matches, $max_offset offset, and $max_error errors\n";}
else { print LOG "Checking for $max_offset meter offset and $max_error errors (std dev of fit)\n";}

if (-e "geo_HH/mk_geo_radcal_2.log") { $log = "geo_HH/mk_geo_radcal_2.log"; }
elsif (-e "geo_hh/mk_geo_radcal_2.log") { $log = "geo_hh/mk_geo_radcal_2.log"; }
elsif (-e "geo_vv/mk_geo_radcal_2.log") { $log = "geo_vv/mk_geo_radcal_2.log"; }
elsif (-e "geo/mk_geo_radcal_2.log") { $log = "geo/mk_geo_radcal_2.log"; }
else { die "ERROR $0: Can't find mk_geo_radcal_2.log file\n"; }

if  (-e $log) {
  open(TLOG,$log) or die "ERROR $0: could not open log: $log: $!\n\n";
  my (@lines) = <TLOG>;
  close(TLOG);

  while (my $line = shift(@lines)) {
 
    #
    # Only checks the number of matches if a values was passed on the command line
    #
    if ($matches_flg != 0) {  
      if ($line =~ /final solution:\s+(\S+)/) {
        $matches = $1;
        print LOG "Number of matches is $matches\n";
        if ($matches < $min_matches) {
          print LOG "WARNING: not enough matches, using dead reckoning\n";
          print LOG "Granule failed coregistration\n";
          $ret = -1;
          close LOG;
          exit $ret;
        } 
      }
    }
 
    if ($line =~ /final range offset poly. coeff.:\s+(\S+)/) {
      $r1 = $1;
      $r2 = 0;
      $r3 = 0;
      $r4 = 0;
      $r5 = 0;
      $r6 = 0;
      print LOG "Range offset is $r1\n";
    }

    if ($line =~ /final range offset poly. coeff.:\s+(\S+)\s+(\S+)\s+(\S+)/) {
      $r1 = $1;
      $r2 = $2;
      $r3 = $3;
      $r4 = 0;
      $r5 = 0;
      $r6 = 0;
      print LOG "Range polynomial is $r1 $r2 $r3\n";
    }

    if ($line =~ /final range offset poly. coeff.:\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)/) {
      $r1 = $1;
      $r2 = $2;
      $r3 = $3;
      $r4 = $4;
      $r5 = $5;
      $r6 = $6;
      print LOG "Range polynomial is $r1 $r2 $r3 $r4 $r5 $r6\n";
    }

    if ($line =~ /final azimuth offset poly. coeff.:\s+(\S+)/) {
      $a1 = $1;
      $a2 = 0;
      $a3 = 0;
      $a4 = 0;
      $a5 = 0; 
      $a6 = 0;
      print LOG "Azimuth offset is $a1\n";
    }

    if ($line =~ /final azimuth offset poly. coeff.:\s+(\S+)\s+(\S+)\s+(\S+)/) {
      $a1 = $1;
      $a2 = $2;
      $a3 = $3;
      $a4 = 0;
      $a5 = 0; 
      $a6 = 0;
      print LOG "Azimuth polynomial is $a1 $a2 $a3\n";
    }
 
    if ($line =~ /final azimuth offset poly. coeff.:\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)/) {
      $a1 = $1;
      $a2 = $2;
      $a3 = $3;
      $a4 = $4; 
      $a5 = $5; 
      $a6 = $6;
      print LOG "Azimuth polynomial is $a1 $a2 $a3 $a4 $a5 $a6\n";
    }
 
    if ($line =~ /final model fit std. dev. \(samples\) range:\s+(\S+)\s+\S+\s+(\S+)/) {
      $rn_error = $1;
      $az_error = $2;
      print LOG "Range std dev: $rn_error  Azimuth std dev: $az_error\n";
      $error = sqrt($rn_error * $rn_error + $az_error * $az_error);
      if ($error > $max_error) {
        print LOG "WARNING: std dev is too high, using dead reckoning\n";
        print LOG "Granule failed coregistration\n";
        $ret = -1;
        close LOG;
        exit $ret; 
      }
    }
  }
}

$log = glob('geo_??/*.diff_par');
if (-e $log) { $platform = "NOT_LEGACY"; }
else {
  $log = glob('geo/*.diff_par');
  if (-e $log) { $platform = "LEGACY"; }
  else { die "ERROR $0: Can't find diff_par file\n"; }
}

if (-e $log) {
  open(TLOG,$log) or die "ERROR $0: could not open log: $log: $!\n\n";
  my (@lines) = <TLOG>;
  close(TLOG);

  while (my $line = shift(@lines)) {
    if ($line =~ /range_samp_1:\s+(\S+)/) {
      $ns = $1;
      print LOG "Number of samples is $ns\n";
    }
    if ($line =~ /az_samp_1:\s+(\S+)/) {
      $nl = $1;
      print LOG "Number of lines is $nl\n";
    }
  }
} else { die "ERROR $0: Can't find diff par file $log\n"; }


# Point1 = 1, 1 
($rpt1,$apt1)=calc(1,1);
$pt1 = sqrt($rpt1*$rpt1 + $apt1*$apt1);
print LOG "Point 1 offset is $pt1\n";

# Point2 = ns, 1
($rpt2,$apt2) = calc($ns,1);
$pt2 = sqrt($rpt2*$rpt2 + $apt2*$apt2);
print LOG "Point 2 offset is $pt2\n";

# Point3 = 1, nl
($rpt3,$apt3) = calc(1,$nl);
$pt3 = sqrt($rpt3*$rpt3 + $apt3*$apt3);
print LOG "Point 3 offset is $pt3\n";

# Point4 = ns, nl
($rpt4,$apt4) = calc($ns,$nl);
$pt4 = sqrt($rpt4*$rpt4 + $apt4*$apt4);
print LOG "Point 4 offset is $pt4\n";

$top = max($pt1, $pt2, $pt3, $pt4);
print LOG "Biggest offset is $top pixels\n";

$offset = $top * $post;
print LOG "Found absolute offset of $offset meters\n";
if ($offset < $max_offset) {
  print LOG "...keeping offset\n";
  $ret = 0;
} else {
  print LOG "WARNING: offset too large, using dead reckoning\n";
  $ret = -1;
}

if ($ret == 0) { print LOG "Granule passed coregistration\n"; }
else { print LOG "Granule failed coregistration\n"; }

close LOG;

exit $ret;

sub max {
    my ($max, @vars) = @_;
    for (@vars) {
        $max = $_ if $_ > $max;
    }
    return $max;
}

sub calc {
    my ($s,$l) = @_;
    
    $rpt1 = $r1 + $r2 * $s + $r3 * $l + $r4*$s*$l + $r5*$s*$s + $r6*$l*$l;
    $apt1 = $a1 + $a2 * $s + $a3 * $l + $a4*$s*$l + $a5*$s*$s + $a6*$l*$l;

    return ($rpt1,$apt1)
}

File: src/check_coreg.py
#!/usr/bin/python

import os, re
import argparse
import logging
import numpy as np
import glob
from getParameter import getParameter

def calc(s,l,r,a):
    rpt = r[0] + r[1]*s + r[2]*l + r[3]*s*l + r[4]*s*s + r[5]*l*l
    apt = a[0] + a[1]*s + a[2]*l + a[3]*s*l + a[4]*s*s + a[5]*l*l
    
    print rpt, apt
    return rpt, apt

def display(str,f,ERROR=False):
    
    if ERROR:
        logging.error("{}".format(str))
    else:
        logging.info("{}".format(str))
    f.write("{}\n".format(str))

def check_coreg(input,post,max_offset=50,max_error=2):

    f = open('coreg_check.log','w')

    display("SAR file: {}".format(input),f)
    display("Checking coregistration using {} meters".format(post),f)
    display("Setting maximum offset to be {}".format(max_offset),f)
    display("Setting maximum error to be {}".format(max_error),f)

    myfile = "mk_geo_radcal_2.log"
    if os.path.isdir("geo_HH"):
        mlog = "geo_HH/{}".format(myfile)
    elif os.path.isdir("geo_hh"):
        mlog = "geo_hh/{}".format(myfile)
    elif os.path.isdir("geo_VV"):
        mlog = "geo_VV/{}".format(myfile)
    elif os.path.isdir("geo_vv"):
        mlog = "geo_vv/{}".format(myfile)
    elif os.path.isdir("geo"):
        mlog = "geo/{}".format(myfile)
    else:
        display("ERROR: Can't find {}".format(myfile),f,ERROR=True)
        f.close()
	exit(-1)

    a = np.zeros(6)
    r = np.zeros(6)
    
    g = open(mlog,"r")
    for line in g:
        if 'final range offset poly. coeff.:' in line:
	    tmp = re.split(":",line)
	    vals = tmp[1].split()
	    if len(vals) == 1:
	        r[0] = float(vals[0])
		display("Range offset is {}".format(r),f)
	    elif len(vals) == 3:
	        r[0] = float(vals[0])
	        r[1] = float(vals[1])
	        r[2] = float(vals[2])
		display("Range polynomial is {}".format(r),f)
            elif len(vals) == 6:
	        r[0] = float(vals[0])
	        r[1] = float(vals[1])
	        r[2] = float(vals[2])
	        r[3] = float(vals[3])
	        r[4] = float(vals[4])
	        r[5] = float(vals[5])
		display("Range polynomial is {}".format(r),f)

	if 'final azimuth offset poly. coeff.:' in line:
	    tmp = re.split(":",line)
	    vals = tmp[1].split()
	    if len(vals) == 1:
	        a[0] = float(vals[0])
		display("Azimuth offset is {}".format(a),f)
	    elif len(vals) == 3:
	        a[0] = float(vals[0])
	        a[1] = float(vals[1])
	        a[2] = float(vals[2])
		display("Azimuth polynomial is {}".format(a),f)
            elif len(vals) == 6:
	        a[0] = float(vals[0])
	        a[1] = float(vals[1])
	        a[2] = float(vals[2])
	        a[3] = float(vals[3])
	        a[4] = float(vals[4])
	        a[5] = float(vals[5])
		display("Azimuth polynomial is {}".format(a),f)
	
        if 'final model fit std. dev. (samples) range:' in line:
	    tmp = re.split(":",line)
	    vals = tmp[1].split()
	    rng_error = float(vals[0])
	    val = tmp[2].strip()
	    azi_error = float(val)
	    display("Range std dev: {}  Azimuth std dev: {}".format(rng_error,azi_error),f)
	    error = np.sqrt(rng_error*rng_error+azi_error*azi_error)
	    display("error is {}".format(error),f)
	    if error > max_error:
	        display("error > max_error",f)
	        display("std dev is too high, using dead reckoning",f)
		display("Granule failed coregistration",f)
                f.close()
		exit(-1)
    g.close()
	
    mlog = glob.glob('geo_??/*.diff_par')[0]
    if mlog:
        platform = "NOT_LEGACY"
    else:
        mlog = glob.glob('geo/*.diff_par')[0]
	if mlog:
	    platform = "LEGACY"
	else:
	    log.error("Can't find diff_par file")
	    return(-1)
    
    if os.path.exists(mlog):
        ns = int(getParameter(mlog,"range_samp_1"))
	display("Number of samples is {}".format(ns),f)
	nl = int(getParameter(mlog,"az_samp_1"))
	display("Number of lines is {}".format(nl),f)
    else:
        log.error("Can't find diff par file {}".format(log))
	f.close()
        exit(-1)
 
    rpt,apt = calc(1,1,r,a)
    pt1 = np.sqrt(rpt*rpt+apt*apt)
    display("Point 1 offset is {}".format(pt1),f)
    
    rpt,apt = calc(ns,1,r,a)
    pt2 = np.sqrt(rpt*rpt+apt*apt)
    display("Point 2 offset is {}".format(pt2),f)
    
    rpt,apt = calc(1,nl,r,a)
    pt3 = np.sqrt(rpt*rpt+apt*apt)
    display("Point 3 offset is {}".format(pt3),f)
    
    rpt,apt = calc(ns,nl,r,a)
    pt4 = np.sqrt(rpt*rpt+apt*apt)
    display("Point 4 offset is {}".format(pt4),f)
    
    top = max(pt1,pt2,pt3,pt4)
    offset = top * post
    
    display("Found absolute offset of {} meters".format(offset),f)
    if offset < max_offset:
        display("...keeping offset",f)
	ret = 0
    else:
        display("offset too large, using dead reckoning",f)
	ret = -1

    if ret == 0:
        display("Granule passed coregistration",f)
	return(0)
    else:
        display("Granule failed coregistration",f)
        f.close()
        exit(-1)


if __name__ == '__main__':

    parser = argparse.ArgumentParser(prog='check_coreg',
      description='Checks results of Gamma RTC coregistration process')
    parser.add_argument('input',help='Name of input SAR file')
    parser.add_argument('post',help='Posting of the SAR image',type=float)
    parser.add_argument('-o','--max_offset',help='Set the maximum allowable max_offset (meters)',type=float,default=50)
    parser.add_argument('-e','--max_error',help='Set the maximum allowable standard deviation of max_offset fit (pixels)',type=int,default=2)
    args = parser.parse_args()

    logFile = "check_coreg.log"
    logging.basicConfig(filename=logFile,format='%(asctime)s - %(levelname)s - %(message)s',
                        datefmt='%m/%d/%Y %I:%M:%S %p',level=logging.DEBUG)
    logging.getLogger().addHandler(logging.StreamHandler())
    logging.info("Starting run")

    check_coreg(args.input,args.post,args.max_offset,args.max_error)
    
